### ISSUE TITLE
Crash in Android: onInitializationComplete is called twice in MobileAds#initialize

### ISSUE THREAD

**Issue by kolotum** [googleads/googleads-mobile-flutter#130]
## Bug report

After we switched from firebase_admob to the google_mobile_ads Plugin (Version: 0.11.0+3) we are seeing repeating crashes of some of our Android users.

`onInitializationComplete()` of `MobileAds#initialize` in `GoogleMobileAdsPlugin.java` is called twice, which is not supported by `DartMessenger`.

Stack trace:

> Fatal Exception: java.lang.IllegalStateException: Reply already submitted
>        at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:155)
>        at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:238)
>        at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin$1.onInitializationComplete(GoogleMobileAdsPlugin.java:211)
>        at com.google.android.gms.internal.ads.zzzn$zza.zze(zzzn.java:6)
>        at com.google.android.gms.internal.ads.zzajn.zza(zzajn.java:11)
>        at com.google.android.gms.internal.ads.zzgy.onTransact(zzgy.java:13)
>        at android.os.Binder.transact(Binder.java:635)
>        at gi.bd(gi.java:2)
>        at com.google.android.gms.ads.internal.initialization.f.e(f.java)
>        at com.google.android.gms.ads.nonagon.initialization.a.run(a.java:2)
>        at android.os.Handler.handleCallback(Handler.java:790)
>        at android.os.Handler.dispatchMessage(Handler.java:99)
>        at qx.a(qx.java)
>        at com.google.android.gms.ads.internal.util.zzaj.a(zzaj.java:1)
>        at qx.dispatchMessage(qx.java)
>        at android.os.Looper.loop(Looper.java:164)
>        at android.app.ActivityThread.main(ActivityThread.java:7000)
>        at java.lang.reflect.Method.invoke(Method.java)
>        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:441)
>        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1408)

### Steps to reproduce

We could not reproduce the crash by ourselves, yet, but we have several affected Android users.

### Expected behavior

No crash, onInitializationComplete called once per app start

---

## Additional context

We use Google-Mobile-Ads as a mediation service and have included additional mediated ad networks:
- AppLovin
- Facebook
- IronSource
- UnityAds
- Vungle

For all affected users the first `InitializationStatus` result of `MobileAds.instance.initialize()` only contains informations about AdMob, the other expected ad networks are missing: 

> adapterStatuses: (com.google.android.gms.ads.MobileAds: AdapterStatus(state: AdapterInitializationState.ready, latency: 0.0))

---

**Comment by lks-nbg** [comment]
It looks like I am having the same issue. In my case the crash so far only occured on old devices and I was able to reproduce it by simply running the example app on an Nexus 5 (Android 6.0.1). 



<details><summary>Click To Expand (logs)</summary>

```
Launching lib\main.dart on Nexus 5 in debug mode...
Running Gradle task 'assembleDebug'...
Note: C:\Dev\flutter\.pub-cache\hosted\pub.dartlang.org\e2e-0.7.0+1\android\src\main\java\dev\flutter\plugins\e2e\E2EPlugin.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: C:\Users\lukas\Desktop\googleads-mobile-flutter-master\packages\google_mobile_ads\example\android\app\src\main\java\io\flutter\plugins\googlemobileadsexample\NativeAdFactoryExample.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
√ Built build\app\outputs\flutter-apk\app-debug.apk.
Installing build\app\outputs\flutter-apk\app.apk...
Debug service listening on ws://127.0.0.1:52383/RZfdGwhXVaA=/ws
Syncing files to device Nexus 5...
I/DynamiteModule(18587): Considering local module com.google.android.gms.ads.dynamite:0 and remote module com.google.android.gms.ads.dynamite:210402101
I/DynamiteModule(18587): Selected remote version of com.google.android.gms.ads.dynamite, version >= 210402101
D/DynamitePackage(18587): Instantiated singleton DynamitePackage.
D/DynamitePackage(18587): Instantiating com.google.android.gms.ads.ChimeraMobileAdsSettingManagerCreatorImpl
I/Ads     (18587): Updating ad debug logging enablement.
I/WebViewFactory(18587): Loading com.google.android.webview version 89.0.4389.105 (code 438910500)
I/cr_LibraryLoader(18587): Loaded native library version number "89.0.4389.105"
I/cr_CachingUmaRecorder(18587): Flushed 4 samples from 4 histograms.
W/cr_media(18587): Requires BLUETOOTH permission
E/libEGL  (18587): validate_display:255 error 3008 (EGL_BAD_DISPLAY)
E/Ads     (18587): Google Mobile Ads SDK initialization functionality unavailable for this session. Ad requests can be made at any time.
E/libEGL  (18587): validate_display:255 error 3008 (EGL_BAD_DISPLAY)
W/Ads     (18587): Invoke Firebase method getInstance error.
W/Ads     (18587): The Google Mobile Ads SDK will not integrate with Firebase. Admob/Firebase integration requires the latest Firebase SDK jar, but Firebase SDK is either missing or out of date
W/Ads     (18587): Update ad debug logging enablement as false
D/AndroidRuntime(18587): Shutting down VM
E/art     (18587): invalid stream - problem with parameter iterator in /data/data/com.google.android.gms/app_chimera/m/0000017b/dl-AdsFdrDynamite.integ_210402101000000.apk for method void com.google.android.gms.ads.internal.initialization.f.e(java.util.List)
E/art     (18587): invalid stream - problem with parameter iterator in /data/data/com.google.android.gms/app_chimera/m/0000017b/dl-AdsFdrDynamite.integ_210402101000000.apk for method void qx.a(android.os.Message)
E/art     (18587): invalid stream - problem with parameter iterator in /data/data/com.google.android.gms/app_chimera/m/0000017b/dl-AdsFdrDynamite.integ_210402101000000.apk for method void qx.dispatchMessage(android.os.Message)
E/AndroidRuntime(18587): FATAL EXCEPTION: main
E/AndroidRuntime(18587): Process: io.flutter.plugins.googlemobileadsexample, PID: 18587
E/AndroidRuntime(18587): java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime(18587): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:155)
E/AndroidRuntime(18587): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:238)
E/AndroidRuntime(18587): 	at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin$1.onInitializationComplete(GoogleMobileAdsPlugin.java:232)
E/AndroidRuntime(18587): 	at com.google.android.gms.internal.ads.zzzn$zza.zze(com.google.android.gms:play-services-ads-lite@@19.7.0:6)
E/AndroidRuntime(18587): 	at com.google.android.gms.internal.ads.zzajn.zza(com.google.android.gms:play-services-ads-lite@@19.7.0:11)
E/AndroidRuntime(18587): 	at com.google.android.gms.internal.ads.zzgy.onTransact(com.google.android.gms:play-services-ads-base@@19.7.0:13)
E/AndroidRuntime(18587): 	at android.os.Binder.transact(Binder.java:387)
E/AndroidRuntime(18587): 	at gi.bd(:com.google.android.gms.policy_ads_fdr_dynamite@210402101@210402101000.356914649.356914649:2)
E/AndroidRuntime(18587): 	at com.google.android.gms.ads.internal.initialization.f.e(:com.google.android.gms.policy_ads_fdr_dynamite@210402101@210402101000.356914649.356914649)
E/AndroidRuntime(18587): 	at com.google.android.gms.ads.nonagon.initialization.a.run(Unknown Source)
E/AndroidRuntime(18587): 	at android.os.Handler.handleCallback(Handler.java:739)
E/AndroidRuntime(18587): 	at android.os.Handler.dispatchMessage(Handler.java:95)
E/AndroidRuntime(18587): 	at qx.a(:com.google.android.gms.policy_ads_fdr_dynamite@210402101@210402101000.356914649.356914649)
E/AndroidRuntime(18587): 	at com.google.android.gms.ads.internal.util.f.a(:com.google.android.gms.policy_ads_fdr_dynamite@210402101@210402101000.356914649.356914649:1)
E/AndroidRuntime(18587): 	at qx.dispatchMessage(:com.google.android.gms.policy_ads_fdr_dynamite@210402101@210402101000.356914649.356914649)
E/AndroidRuntime(18587): 	at android.os.Looper.loop(Looper.java:148)
E/AndroidRuntime(18587): 	at android.app.ActivityThread.main(ActivityThread.java:5417)
E/AndroidRuntime(18587): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(18587): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
E/AndroidRuntime(18587): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
I/flutter (18587): Initialization done: {com.google.android.gms.ads.MobileAds: Instance of 'AdapterStatus'}
I/Process (18587): Sending signal. PID: 18587 SIG: 9
Lost connection to device.

```

</details>

***

**Describe the bug**
As mentioned above, the app crashes immediately on start-up.

### Steps to reproduce
Run the example app from google_mobile_ads 0.12.0 on an old device, probably Android 6.0.1 and older versions (assuming the issue #95 is related to this).

### Expected behavior
No crash

---

## Additional context
• Nexus 5 (mobile)
• android-arm
• Android 6.0.1 (API 23)

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 2.0.3, on Microsoft Windows [Version 10.0.19041.867], locale de-DE)
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[√] Android Studio (version 4.1.0)
[√] Connected device (4 available)

• No issues found!

```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
Dart SDK 2.12.2
Flutter SDK 2.0.3
google_mobile_ads_example 0.0.0

dependencies:
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- google_mobile_ads 0.12.0 [meta flutter]

dev dependencies:
- e2e 0.7.0+1 [flutter flutter_driver flutter_test path]
- flutter_driver 0.0.0 [file meta path vm_service webdriver flutter flutter_test fuchsia_remote_debug_protocol archive args async boolean_selector characters charcode clock collection convert crypto fake_async matcher platform process source_span stack_trace stream
_channel string_scanner sync_http term_glyph test_api typed_data vector_math]
- pedantic 1.11.0

transitive dependencies:
- archive 2.0.13 [crypto args path]
- args 1.6.0
- async 2.5.0 [collection]
- boolean_selector 2.1.0 [source_span string_scanner]
- characters 1.1.0
- charcode 1.2.0
- clock 1.1.0
- collection 1.15.0
- convert 2.1.1 [charcode typed_data]
- crypto 2.1.5 [collection convert typed_data]
- fake_async 1.2.0 [clock collection]
- file 6.0.0 [meta path]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]
- fuchsia_remote_debug_protocol 0.0.0 [process vm_service file meta path platform]
- matcher 0.12.10 [stack_trace]
- meta 1.3.0
- path 1.8.0
- platform 3.0.0
- process 4.0.0 [file path platform]
- sky_engine 0.0.99
- source_span 1.8.0 [charcode collection path term_glyph]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- string_scanner 1.1.0 [charcode source_span]
- sync_http 0.2.0
- term_glyph 1.2.0
- test_api 0.2.19 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- typed_data 1.3.0 [collection]
- vector_math 2.1.0
- vm_service 5.5.0 [meta]
- webdriver 2.1.2 [archive matcher path stack_trace sync_http]
```

</details>

---

**Comment by greymag** [comment]
Any updates on this? This is pretty critical issue - some user just can't use app at all.

**Comment by kolotum** [comment]
> Any updates on this? This is pretty critical issue - some user just can't use app at all.

We have added a workaround in our fork: https://github.com/lotum/googleads-mobile-flutter/commit/868811eab40800b12e1a3da2afe1317ee6ff424f

We abort when onInitializationComplete is called again. So far we are not seeing any issues using this solution and the crash does not occur anymore with the above workaround.

**Comment by guenth39** [comment]
I have the same problem, but the error message shows different lines. I don't know if this has anything to do with anything, but I wanted to share it with you. The device is an LG V30 with Android 8.0.

```
java.lang.IllegalStateException: Reply already submitted
	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(Unknown Source:35)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(Unknown Source:14)
	at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin$1.onInitializationComplete(Unknown Source:7)
	at com.google.android.gms.internal.ads.Ba.a(Unknown Source:2)
	at com.google.android.gms.internal.ads.xa.run(Unknown Source:4)
	at android.os.Handler.handleCallback(Handler.java:789)
	at android.os.Handler.dispatchMessage(Handler.java:98)
	at com.google.android.gms.internal.ads.aaa.a(Unknown Source:0)
	at com.google.android.gms.internal.ads.aaa.dispatchMessage(Unknown Source:0)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6703)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:769)
```

I have tried @kolotum s fix but unfortunately today I got the feedback from a user that it still does not work. The error message is the same, but the app seems to crash right away and not after 1-2 seconds.

Unfortunately, I can't reproduce the error on any of my devices. Has anyone found a way to do this? Then it would certainly be easier to find a solution.

Best regards

**Comment by jjliu15** [comment]
Should be fixed in latest release: v0.13.2+1
